slatedb/src/compactor.rs

```diff
@@ -876,6 +898,530 @@ mod tests {
+                RowEntry::new_value(&[b'y'; 16], &[b'p'; 128], 6).with_create_ts(0),
+            ],
+        )
+        .await;
+
+        // verify the merged value is readable from the db
+        let result = db.get(b"key1").await.unwrap();
+        assert_eq!(result, Some(Bytes::from("abcd")));
+    }
+
+    #[tokio::test(flavor = "multi_thread", worker_threads = 2)]
+    async fn should_merge_without_base_value() {
+        use crate::test_utils::OnDemandCompactionSchedulerSupplier;
+
+        // given:
+        let os = Arc::new(InMemory::new());
+        let logical_clock = Arc::new(TestClock::new());
+        let compaction_scheduler = Arc::new(OnDemandCompactionSchedulerSupplier::new(Arc::new(
+            |state| state.db_state().l0.len() >= 2,
+        )));
+        let options = db_options(Some(compactor_options()));
+
+        let db = Db::builder(PATH, os.clone())
+            .with_settings(options)
+            .with_logical_clock(logical_clock)
+            .with_compaction_scheduler_supplier(compaction_scheduler)
+            .with_merge_operator(Arc::new(StringConcatMergeOperator))
+            .build()
+            .await
+            .unwrap();
+
+        let (_manifest_store, table_store) = build_test_stores(os.clone());
+
+        // write only merge operations without any base value
+        db.merge(b"key1", b"x").await.unwrap();
+        db.put(&vec![b'x'; 16], &vec![b'p'; 128]).await.unwrap(); // padding to exceed 256 bytes
+        db.flush().await.unwrap();
+
+        db.merge(b"key1", b"y").await.unwrap();
+        db.merge(b"key1", b"z").await.unwrap();
+        db.put(&vec![b'y'; 16], &vec![b'p'; 128]).await.unwrap(); // padding to exceed 256 bytes
+        db.flush().await.unwrap();
+
+        // when:
+        let db_state = await_compaction(&db).await;
+
+        // then:
+        let db_state = db_state.expect("db was not compacted");
+        assert_eq!(db_state.compacted.len(), 1);
+        let compacted = &db_state.compacted.first().unwrap().ssts;
+        assert_eq!(compacted.len(), 1);
+        let handle = compacted.first().unwrap();
+
+        let mut iter = SstIterator::new_borrowed_initialized(
+            ..,
+            handle,
+            table_store.clone(),
+            SstIteratorOptions::default(),
+        )
+        .await
+        .unwrap()
+        .expect("Expected Some(iter) but got None");
+
+        assert_iterator(
+            &mut iter,
+            vec![
+                RowEntry::new_merge(b"key1", b"xyz", 4).with_create_ts(0),
+                RowEntry::new_value(&[b'x'; 16], &[b'p'; 128], 2).with_create_ts(0),
+                RowEntry::new_value(&[b'y'; 16], &[b'p'; 128], 5).with_create_ts(0),
+            ],
+        )
+        .await;
+
+        // verify the merged value is readable from the db
+        let result = db.get(b"key1").await.unwrap();
+        assert_eq!(result, Some(Bytes::from("xyz")));
+    }
+
```

Comment:
should we add a test to ensure we don't merge across `put`s? E.g. a sequence like:
```
put("x", "a");
merge("x", "b");
put("x", "c");
merge("x", "d");
```

Should return 2 entries for x - one with the value "ab" and the next with the value "cd". It's important for snapshot-reads.

Reply:
@rodesai I went to add this test but we currently hardcode the compaction logic of only keeping the latest value. This applies to both merges and normal puts (i.e. two normal puts will result in only one key). The first `None` indicates that we should not keep old version of a key.

https://github.com/slatedb/slatedb/blob/d4f75d51ca946de91a858bdd266a8e27f76de3a0/slatedb/src/compactor_executor.rs#L195-L203

cc @flaneur2020 to make sure I get this right